File: Semester 3/Class 3 Problems/YearOfTheCow.java
import java.util.*;
import java.io.*;

public class YearOfTheCow {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static PrintWriter pr = new PrintWriter(new BufferedWriter(new OutputStreamWriter(System.out)));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        String[] calender = { "Ox", "Tiger", "Rabbit", "Dragon", "Snake",
                "Horse", "Goat", "Monkey", "Rooster", "Dog", "Pig", "Rat" };

        int n = readInt();
        HashMap<String, Integer> years = new HashMap<>();
        for (int i = 1; i <= calender.length; i++) {
            years.put(calender[i - 1], i);
        }
        HashMap<String, Integer> map = new HashMap<>();
        HashMap<String, String> chineseYear = new HashMap<>();

        for (int i = 0; i < n; i++) {
            String[] line = readLine().split(" ");
            String name = line[0];
            chineseYear.put(name, line[4]);
            if (line[7].equals("Bessie")) {

                int year = 0;
                if (line[3].equals("previous")) {
                    year = years.get(line[4]) - 13;
                } else {
                    year = years.get(line[4]) - 1;
                    if (year == 0)
                        year = 12;
                }
                map.put(name, year);
            } else {
                String previousName = line[7];
                int previousYear = map.get(previousName);
                int year = 0;

                String c_year = chineseYear.get(previousName);
                Integer c_year_n = years.get(c_year);
                int y = years.get(line[4]);
                if (line[3].equals("previous")) {

                    if (y > c_year_n) {
                        year = (12 + c_year_n - y) * -1;
                    } else if (y == c_year_n) {
                        year = -12;
                    } else {
                        year = y - c_year_n;
                    }

                } else {
                    if (y > c_year_n) {
                        year = y - c_year_n;
                    } else {
                        year = 12 + y - c_year_n;
                    }
                }
                year += previousYear;
                map.put(name, year);
                if (name.equals("Elsie")) {
                    System.out.println(Math.abs(map.get(name)));
                }
            }
        }
    }

    static String next() throws IOException {
        while (st == null || !st.hasMoreTokens())
            st = new StringTokenizer(br.readLine().trim());
        return st.nextToken();
    }

    static long readLong() throws IOException {
        return Long.parseLong(next());
    }

    static int readInt() throws IOException {
        return Integer.parseInt(next());
    }

    static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }

    static char readCharacter() throws IOException {
        return next().charAt(0);
    }

    static String readLine() throws IOException {
        return br.readLine().trim();
    }
}

File: Semester 4/Class13_Computer_Contest_Level_3_Notes_and_Homework_Dec_03/lvl3_hw_sol/year_of_the_cow.cpp
#include <bits/stdc++.h>
using namespace std;
typedef pair<int, int> pi;
vector<string> zodiac ={"Ox", "Tiger", "Rabbit", "Dragon", "Snake", "Horse", "Goat", "Monkey", "Rooster", "Dog", "Pig", "Rat"};
unordered_map<string, pi> cow;
int n; vector<string> s(8);
int main(){
    cin >> n; cow["Bessie"] = {0, 0};  //{zodiac, age}
    for(int k=1; k<=n; k++){
        for(int i=0; i<s.size(); i++) cin >> s[i];
        string ncow = s[0], ocow = s[7];
        int pos = find(zodiac.begin(), zodiac.end(), s[4]) - zodiac.begin(), dif = 0;
        int opos = cow[ocow].first, oage = cow[ocow].second;
        if(s[3] == "previous") {
            if(opos > pos) dif = opos - pos;
            else dif = 12 + opos - pos;
            cow[ncow] = {pos, oage + dif};
            //cout << ncow << " " << zodiac[pos] << " " << oage + dif << endl;
        }else {
            if(opos >= pos) dif = 12 + pos - opos;
            else dif = pos - opos;
            cow[ncow] = {pos, oage - dif};
            //cout << ncow << " " << zodiac[pos] << " " << oage - dif << endl;
        }
    }
    cout << abs(cow["Bessie"].second - cow["Elsie"].second) << endl;
}